src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
     // Ensure that of 2 accesses happen, one foreign and one a child, and we are protected, that we
     // get UB unless they are both reads.
     fn protected_enforces_noalias() {
-        for rel1 in AccessRelatedness::all() {
-            for rel2 in AccessRelatedness::all() {
-                if rel1.is_foreign() == rel2.is_foreign() {
-                    // We want to check pairs of accesses where one is foreign and one is not.
-                    continue;
+        for [rel1, rel2] in <[AccessRelatedness; 2]>::exhaustive() {
+            // We want to check pairs of accesses where one is foreign and one is not.
+            precondition!(rel1.is_foreign() != rel2.is_foreign());
+            for [kind1, kind2] in <[AccessKind; 2]>::exhaustive() {
+                for mut state in LocationState::exhaustive() {
+                    let protected = true;
+                    precondition!(state.perform_access(kind1, rel1, protected).is_ok());
+                    precondition!(state.perform_access(kind2, rel2, protected).is_ok());
+                    // If these were both allowed, it must have been two reads.
+                    assert!(
+                        kind1 == AccessKind::Read && kind2 == AccessKind::Read,
+                        "failed to enforce noalias between two accesses that are not both reads"
+                    );
+                }
+            }
+        }
+    }
+
+    // Possible pairs of `AccessRelatedness` that can occur
```

Comment:
Please move these tests into a new file... this almost double the current file size.^^
(You can make it `tree/tests.rs` or so if you need access to private stuff of `tree`.)